Implement `read_buf` and vectored read/write for SGX stdio

thaliaarchi opened the pull request:
Implement `read_buf`, `read_vectored`, and `write_vectored` for the SGX stdio types.

Additionally, extend `User<T>::copy_to_enclave` to work for copying to uninitialized values and fix unsoundness in `UserRef<[T]>::copy_to_enclave_vec`.

cc @jethrogb

Tracked in https://github.com/rust-lang/rust/issues/136756

thaliaarchi committed:
Implement read_buf and vectored read/write for SGX stdio

thaliaarchi commented:
@bors blocked (on https://github.com/rust-lang/rust/pull/136780)

jethrogb commented:
cc @raoulstrackx

jethrogb commented on library/std/src/sys/pal/sgx/abi/usercalls/alloc.rs:
```diff
@@ -651,13 +651,29 @@ where
+    /// # Panics
+    /// This function panics if the destination doesn't have the same length as
+    /// the source.
+    pub fn copy_to_enclave_uninit(&self, dest: &mut [MaybeUninit<T>]) {
```
Needing this “feels” wrong to me. You could do `impl<T: UserSafeSized> UserSafeSized for MaybeUninit<T>` and then just create a `[MaybeUninit<u8>]` in `read_buf`. However, of course the whole point of `UserSafe` is that you want to explicitly assume everything you copy out of userspace is initialized, so that would kind of defeat the point.

thaliaarchi replied:
If both of those options are unsatisfying to you, what would you recommend I do so this can impl `read_buf`?

jethrogb commented:
> The versions of `read` and `write` in usercalls are vectored
This was [explicitly requested](https://github.com/rust-lang/rust/pull/59869#discussion_r274244624) by the T-libs reviewer.

thaliaarchi commented:
> > The versions of `read` and `write` in usercalls are vectored
> 
> This was [explicitly requested](https://github.com/rust-lang/rust/pull/59869#discussion_r274244624) by the T-libs reviewer.

I'm fine with reverting my change to this. Since `new_from_enclave` just defers to `new_uninit_bytes` and `copy_from_enclave`, which is what the vectored version uses, the only difference is saving a few arithmetic ops.

thaliaarchi committed:
Fix `UserRef<[T]>::copy_to_enclave_vec`

It reinterprets uninitialized memory as initialized and does not drop
existing elements of the Vec. Fix that.

Additionally, make it more general by appending, instead of overwriting
existing elements, and rename it to `append_to_enclave_vec`. A caller
can simply call `.clear()` before, for the old behavior.

thaliaarchi commented:
I noticed that `UserRef<[T]>::copy_to_enclave_vec` reinterprets uninitialized memory as initialized and does not drop existing elements of the Vec, so I pushed a fix for it which uses the new `copy_to_enclave_uninit`. I think this supports the need for such a routine.

jethrogb commented:
Perhaps `copy_to_enclave` needs to be replaced in its entirety with your version?

thaliaarchi commented:
> Perhaps `copy_to_enclave` needs to be replaced in its entirety with your version?

I think the two would actually be better coexisting, since they serve slightly different purposes. `copy_to_enclave` copies from a `UserRef<T>` to a `&mut T` and `copy_to_enclave_uninit` copies from a `UserRef<[T]>` to a `&mut [MaybeUninit<T>]`. Since `MaybeUninit<T: Sized>`, a routine which copies from a `UserRef<T>` to a `&mut MaybeUninit<T>` wouldn't work for slices.

jethrogb commented:
How about this?
```rust

trait UserSafeCopyDestination<T: ?Sized> {
    fn as_mut_ptr(&mut self) -> *mut T;
}

impl<T> UserSafeCopyDestination<T> for T {
    fn as_mut_ptr(&mut self) -> *mut T {
        self as _
    }
}

impl<T> UserSafeCopyDestination<[T]> for [T] {
    fn as_mut_ptr(&mut self) -> *mut [T] {
        self as _
    }
}

impl<T> UserSafeCopyDestination<T> for MaybeUninit<T> {
    fn as_mut_ptr(&mut self) -> *mut T {
        self as *mut Self as _
    }
}

impl<T> UserSafeCopyDestination<[T]> for [MaybeUninit<T>] {
    fn as_mut_ptr(&mut self) -> *mut [T] {
        self as *mut Self as _
    }
}

impl<T: ?Sized> UserRef<T> {
    pub fn copy_to_enclave<V: ?Sized + UserSafeCopyDestination<T>>(&self, dest: &mut V) {
        unsafe {
            assert_eq!(mem::size_of_val(dest), mem::size_of_val(&*self.0.get()));
            copy_from_userspace(
                self.0.get() as *const T as *const u8,
                dest.as_mut_ptr() as *mut u8,
                mem::size_of_val(dest),
            );
        }
    }
}
```

Commit:
Copy from userspace to MaybeUninit

Co-authored-by: Thalia Archibald <thalia@archibald.dev>

thaliaarchi commented:
That's much more flexible! I've added a commit with your patch, which I attributed to you.

(FYI, it looks like your fortanix.com email isn't connected to your GitHub account, so you're not linked.)

thaliaarchi commented:
Now that https://github.com/rust-lang/rust/pull/136780 has merged, we're good to go!

@rustbot review

jethrogb commented:
> And since we've been talking about SGX copying APIs, I think UserRef<[T]>::copy_to_enclave_vec would be slightly better if it didn't clear existing elements and instead just appended. It's easy for the caller to .clear(), if they want, and this seems to better match other APIs that take a &mut Vec<T>.

Good idea. But then let's also change the name to better describe the behavior.

thaliaarchi commented:
I've now done so and renamed it to `append_to_enclave_vec`.

ChrisDenton commented:
r=me if @jethrogb is happy with this PR.

jethrogb commented:
@ChrisDenton yes please approve

ChrisDenton commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit e80df9c088b91c7efd94d5cce7fc90e202d06ec6 has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton e80df9c088b91c7efd94d5cce7fc90e202d06ec6 -->
<!-- homu: {"type":"Approved","sha":"e80df9c088b91c7efd94d5cce7fc90e202d06ec6","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

thaliaarchi commented:
I had neglected to update the `set_len` for the append-style logic. I've pushed a fix for that (so it needs a reapproval).

ChrisDenton commented:
@bors r-

ChrisDenton commented:
@bors r+

bors commented:
:pushpin: Commit c62aa0baa1a8228d5bfbb3e810db4c7ee77eb3a1 has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton c62aa0baa1a8228d5bfbb3e810db4c7ee77eb3a1 -->
<!-- homu: {"type":"Approved","sha":"c62aa0baa1a8228d5bfbb3e810db4c7ee77eb3a1","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
